internal/xds/rbac/rbac_engine.go

```diff
@@ -15,68 +15,218 @@
+// decisions on incoming RPCs.
+type ChainEngine struct {
+	chainedEngines []*engine
+}
+
+// NewChainEngine returns a chain of RBAC engines, used to make authorization
+// decisions on incoming RPCs. Returns a non-nil error for invalid policies.
+func NewChainEngine(policies []*v3rbacpb.RBAC) (*ChainEngine, error) {
+	var engines []*engine
+	for _, policy := range policies {
+		engine, err := newEngine(policy)
+		if err != nil {
+			return nil, err
+		}
+		engines = append(engines, engine)
+	}
+	return &ChainEngine{chainedEngines: engines}, nil
+}
+
+// IsAuthorized determines if an incoming RPC is authorized based on the chain of RBAC
+// engines and their associated actions.
+//
+// Errors returned by this function are compatible with the status package.
+func (cre *ChainEngine) IsAuthorized(ctx context.Context) error {
+	// This conversion step (i.e. pulling things out of ctx) can be done once,
+	// and then be used for the whole chain of RBAC Engines.
+	rpcData, err := newRPCData(ctx)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "missing fields in ctx %+v: %v", ctx, err)
+	}
+	for _, engine := range cre.chainedEngines {
+		// TODO: What do I do now with this matchingPolicyName?
+		_, err := engine.findMatchingPolicy(rpcData)
+
+		switch {
+		case engine.action == allow && err == errPolicyNotFound:
+			return status.Errorf(codes.PermissionDenied, "incoming RPC %v did not match an allow policy", rpcData)
+		case engine.action == deny && err != errPolicyNotFound:
+			return status.Errorf(codes.PermissionDenied, "incoming RPC %+v matched a deny policy", rpcData)
+		}
+	}
+	// If the incoming RPC gets through all of the engines successfully (i.e.
+	// doesn't not match an allow or match a deny engine), the RPC is authorized
+	// to proceed.
+	return status.Error(codes.OK, "")
+}
+
+type action int
+
+const (
+	allow action = iota
+	deny
+)
+
+// engine is used for matching incoming RPCs to policies.
+type engine struct {
 	policies map[string]*policyMatcher
+	action   action
 }
 
-// NewEngine creates an RBAC Engine based on the contents of policy. If the
-// config is invalid (and fails to build underlying tree of matchers), NewEngine
-// will return an error. This created RBAC Engine will not persist the action
-// present in the policy, and will leave up to caller to handle the action that
-// is attached to the config.
-func NewEngine(policy *v3rbacpb.RBAC) (*Engine, error) {
-	policies := make(map[string]*policyMatcher)
+// newEngine creates an RBAC Engine based on the contents of policy. Returns a
+// non-nil error if the policy is invalid.
+func newEngine(policy *v3rbacpb.RBAC) (*engine, error) {
+	var action action
+	switch *policy.Action.Enum() {
+	case v3rbacpb.RBAC_ALLOW:
+		action = allow
+	case v3rbacpb.RBAC_DENY:
+		action = deny
+	case v3rbacpb.RBAC_LOG:
+		return nil, errors.New("unsupported action")
+	}
+
+	policies := make(map[string]*policyMatcher, len(policy.Policies))
```

Comment:
The envoy docs refer to it as `RBAC configuration`. I guess `config` or `rbacConfig` would be a good name. I would prefer the former since we are in the `rbac` package, so the latter would be a little redundant.